File: apps/frontend/src/types/invoice.ts
export type InvoiceStatus =
  | 'draft'
  | 'ready'
  | 'sent'
  | 'partial_paid'
  | 'paid'
  | 'overdue'
  | 'void'

export type InvoiceLineType =
  | 'rent'
  | 'deposit'
  | 'utility_deposit'
  | 'subscription_utility'
  | 'metered_utility'
  | 'credit'
  | 'adjustment'
  | 'carried_forward'

export type LineItemStatus = 'confirmed' | 'pending'

export interface TierBand {
  band: string
  units: number
  rate: number
  subtotal: number
}

export interface InvoiceLineItem {
  id: string
  type: InvoiceLineType
  description: string
  utility_key?: string
  quantity: number
  unit_price: number
  amount: number
  meter_ticket_id?: string
  current_reading?: number
  previous_reading?: number
  meter_image_url?: string
  status: LineItemStatus
  tier_breakdown?: TierBand[]  // computed band-by-band breakdown (confirmed metered only)
}

export interface SmartMeterDailyReading {
  date: string
  readings: number
  total_usage: number
  avg_usage: number
}

export interface SmartMeterSummary {
  unit_id: string
  utility_key: string
  reading_count: number
  total_usage: number        // total L consumed
  avg_daily_usage: number    // avg L/day
  peak_day: string
  peak_usage: number
  trend_pct: number
  trend_direction: 'up' | 'down' | 'stable'
  advice: string[]
  daily_breakdown: SmartMeterDailyReading[]
  latest_reading: number
  previous_reading: number
  applied_at: string
}

export interface Invoice {
  id: string
  org_id: string
  property_id: string
  unit_id: string
  lease_id: string
  tenant_id: string
  idempotency_key: string
  billing_month: string
  invoice_category: 'rent' | 'deposit'
  status: InvoiceStatus
  sandbox: boolean
  reference_no: string
  due_date: string
  line_items: InvoiceLineItem[]
  subtotal: number
  tax_amount: number
  total_amount: number
  amount_paid: number
  balance_due: number
  carried_forward: number
  notes?: string
  sent_at?: string
  paid_at?: string
  created_by?: string
  created_at: string
  updated_at: string
  smart_meter_summary?: SmartMeterSummary
  // Enriched
  tenant_name?: string
  property_name?: string
  unit_label?: string
}

export interface InvoiceListResponse {
  items: Invoice[]
  total: number
  page: number
  page_size: number
}

export interface InvoiceCounts {
  draft: number
  ready: number
  sent: number
  partial_paid: number
  paid: number
  overdue: number
  void: number
  total: number
}

export interface BillingCycleRun {
  id: string
  org_id: string
  billing_month: string
  run_type: string
  sandbox: boolean
  triggered_by?: string
  status: string
  invoices_created: number
  invoices_skipped: number
  invoices_failed: number
  dry_run_preview?: Record<string, unknown>[]
  failures: Record<string, unknown>[]
  started_at: string
  completed_at?: string
  created_at: string
  updated_at: string
}

export interface BillingRunListResponse {
  items: BillingCycleRun[]
  total: number
  page: number
  page_size: number
}

export interface InvoiceGeneratePayload {
  billing_month: string
  sandbox: boolean
  dry_run: boolean
}

/** Returned by POST /invoices/generate when dry_run=false (async path). */
export interface BillingRunTriggerResponse {
  job_id: string
  status: 'queued'
  billing_month: string
}

export type GenerateBillingResponse = BillingCycleRun | BillingRunTriggerResponse

/** True when the generate response is an async queued trigger (not a dry-run result). */
export function isBillingRunQueued(r: GenerateBillingResponse): r is BillingRunTriggerResponse {
  return 'job_id' in r && !('invoices_created' in r)
}

export interface InvoiceUpdatePayload {
  status?: InvoiceStatus
  notes?: string
  due_date?: string
}

export interface InvoicePaymentPayload {
  amount: number
  method: string
  payment_date: string
  reference?: string
  notes?: string
}

File: apps/backend/app/schemas/invoice.py
from datetime import date, datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel


class TierBandResponse(BaseModel):
    band: str
    units: float
    rate: float
    subtotal: float


class InvoiceLineItemResponse(BaseModel):
    id: str
    type: str
    description: str
    utility_key: Optional[str] = None
    quantity: float
    unit_price: float
    amount: float
    meter_ticket_id: Optional[str] = None
    current_reading: Optional[float] = None
    previous_reading: Optional[float] = None
    meter_image_url: Optional[str] = None  # presigned URL
    status: str
    tier_breakdown: Optional[List["TierBandResponse"]] = None


class InvoiceResponse(BaseModel):
    id: str
    org_id: str
    property_id: str
    unit_id: str
    lease_id: str
    tenant_id: str
    idempotency_key: str
    billing_month: str
    invoice_category: str = "rent"
    status: str
    sandbox: bool
    reference_no: str
    due_date: date
    line_items: List[InvoiceLineItemResponse] = []
    subtotal: float
    tax_amount: float
    total_amount: float
    amount_paid: float
    balance_due: float
    carried_forward: float
    notes: Optional[str] = None
    sent_at: Optional[datetime] = None
    paid_at: Optional[datetime] = None
    created_by: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    smart_meter_summary: Optional[Dict[str, Any]] = None

    # Enriched fields
    tenant_name: Optional[str] = None
    property_name: Optional[str] = None
    unit_label: Optional[str] = None


class InvoiceListResponse(BaseModel):
    items: List[InvoiceResponse]
    total: int
    page: int
    page_size: int


class InvoiceCountsResponse(BaseModel):
    draft: int = 0
    ready: int = 0
    sent: int = 0
    partial_paid: int = 0
    paid: int = 0
    overdue: int = 0
    void: int = 0
    total: int = 0


class InvoiceGenerateRequest(BaseModel):
    billing_month: str          # "YYYY-MM"
    sandbox: bool = False
    dry_run: bool = False


class InvoiceUpdateRequest(BaseModel):
    status: Optional[str] = None
    notes: Optional[str] = None
    due_date: Optional[date] = None


class InvoicePaymentRequest(BaseModel):
    amount: float
    method: str                 # cash | bank_transfer | mpesa_stk | manual
    payment_date: date
    reference: Optional[str] = None
    notes: Optional[str] = None


class BillingCycleRunResponse(BaseModel):
    id: str
    org_id: str
    billing_month: str
    run_type: str
    sandbox: bool
    triggered_by: Optional[str] = None
    status: str
    invoices_created: int
    invoices_skipped: int
    invoices_failed: int
    dry_run_preview: Optional[List[Dict[str, Any]]] = None
    failures: List[dict] = []
    started_at: datetime
    completed_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime


class BillingCycleRunListResponse(BaseModel):
    items: List[BillingCycleRunResponse]
    total: int
    page: int
    page_size: int


class BillingRunTriggerResponse(BaseModel):
    """Returned by POST /invoices/generate when dry_run=False (async path)."""
    job_id: str
    status: str          # "queued"
    billing_month: str
